airflow/contrib/hooks/gcs_hook.py

```diff
@@ -564,6 +564,167 @@ def compose(self, bucket_name, source_objects, destination_object):
+            destination. By default extra files are not deleted.
+
+            .. note::
+                This option can delete data quickly if you specify the wrong source/destination combination.
+
+        :type delete_extra_files: bool
+        :return: none
+        """
+        client = self.get_conn()
+        # Create bucket object
+        source_bucket_obj = client.bucket(source_bucket)
+        destination_bucket_obj = client.bucket(destination_bucket)
+        # Normalize parameters when they are passed
+        source_object = (
+            source_object + "/" if source_object and not source_object.endswith("/") else source_object
+        )
+        destination_object = (
+            destination_object + "/"
+            if destination_object and not destination_object.endswith("/")
+            else destination_object
+        )
+        # Calculate the number of characters that remove from the name, because they contain information
+        # about the parent's path
+        source_object_prefix_len = len(source_object) if source_object else 0
+        # Prepare synchronization plan
+        to_copy_blobs, to_delete_blobs, to_rewrite_blobs = self._prepare_sync_plan(
+            source_bucket=source_bucket_obj,
+            destination_bucket=destination_bucket_obj,
+            source_object=source_object,
+            destination_object=destination_object,
+            recursive=recursive
+        )
+        self.log.info(
+            "Planned synchronization. To delete blobs count: %s, to upload blobs count: %s, "
+            "to rewrite blobs count: %s",
+            len(to_delete_blobs),
+            len(to_copy_blobs),
+            len(to_rewrite_blobs),
+        )
+
+        # Copy missing object to new bucket
+        if to_copy_blobs:
+            for blob in to_copy_blobs:
+                dst_object = (
+                    path.join(destination_object, blob.name[source_object_prefix_len:])
+                    if destination_object
+                    else blob.name[source_object_prefix_len:]
+                )
+                self.copy(
+                    source_bucket=source_bucket_obj.name,
+                    source_object=blob.name,
+                    destination_bucket=destination_bucket_obj.name,
+                    destination_object=dst_object,
+                )
+            self.log.info("Blobs copied.")
+        # Delete redundant files
+        if to_delete_blobs:
+            if delete_extra_files:
+                # TODO: Add batch. I tried to do it, but the Google library is not stable at the moment.
+                for blob in to_delete_blobs:
+                    self.delete(blob.bucket.name, blob.name)
+                self.log.info("Blobs deleted.")
+            else:
+                self.log.info("Skipped blobs deleting")
+
+        # Overwrite files that are different
+        if to_rewrite_blobs:
+            if allow_overwrite:
```

Comment:
This could be done with only one indented block:
```python
if not to_rewrite_blobs:
    self.log.info("Skipped blobs overwriting")
elif allow_overwrite:
    for blob in to_rewrite_blobs:
        do_something()
```
Same above, WDYT?

Reply:
This is slightly different behavior.
If there are no files to overwrite, there should always  do nothing.
if there are files, then should print the appropriate message.

Reply:
Oh I see, but still I think it this should be a binary information: rewriting / skipping. Because otherwise user see only 2 messages that describes 3 cases.

Reply:
I updated this PR.